src/poly/evaluator.rs

```diff
@@ -0,0 +1,386 @@
+    /// _new_ rotation. Existing rotations applied to this leaf node are ignored and the
+    /// returned polynomial is not rotated _relative_ to the previous structure.
+    pub(crate) fn with_rotation(&self, rotation: Rotation) -> Self {
+        AstLeaf {
+            id: self.id,
+            rotation,
+            _evaluator: PhantomData::default(),
+        }
+    }
+}
+
+/// An evaluation context for polynomial operations.
+///
+/// This context enables us to de-duplicate queries of circuit columns (and the rotations
+/// they might require), by storing a list of all the underlying polynomials involved in
+/// any query (which are almost certainly column polynomials). We use the context like so:
+///
+/// - We register each underlying polynomial with the evaluator, which returns a reference
+///   to it as a [`AstLeaf`].
+/// - The references are then used to build up a [`Ast`] that represents the overall
+///   operations to be applied to the polynomials.
+/// - Finally, we call [`Evaluator::evaluate`] passing in the [`Ast`].
+pub(crate) struct Evaluator<E, Id: Eq + Hash, F: Field, B: Basis> {
+    polys: HashMap<Id, Polynomial<F, B>>,
+    _context: E,
+}
+
+/// Constructs a new `Evaluator`.
+///
+/// The `context` parameter is used to provide type safety for evaluators. It ensures that
+/// an evaluator will only be used to evaluate [`Ast`]s containing [`AstLeaf`]s obtained
+/// from itself. It should be set to the empty closure `|| {}`, because anonymous closures
+/// all have unique types.
+pub(crate) fn new_evaluator<E: Fn() + Clone, Id: Eq + Hash, F: Field, B: Basis>(
+    context: E,
+) -> Evaluator<E, Id, F, B> {
+    Evaluator {
+        polys: HashMap::default(),
+        _context: context,
+    }
+}
+
+impl<E, F: Field, B: Basis> Evaluator<E, usize, F, B> {
+    /// Registers the given polynomial for use in this evaluation context.
+    ///
+    /// This API treats each registered polynomial as unique, even if the same polynomial
+    /// is added multiple times.
+    pub(crate) fn register_poly(&mut self, poly: Polynomial<F, B>) -> AstLeaf<E, usize, B> {
+        let id = self.polys.len();
+        self.polys.insert(id, poly);
+
+        AstLeaf {
+            id,
+            rotation: Rotation::cur(),
+            _evaluator: PhantomData::default(),
+        }
+    }
+}
+
+impl<E, Id: Copy + Eq + Hash, F: Field, B: Basis> Evaluator<E, Id, F, B> {
+    /// Registers the given polynomial for use in this evaluation context.
+    ///
+    /// This API registers the polynomial with the given identifier, as a cheaper way of
+    /// detecting duplicate polynomials than an equality check. The caller must ensure the
+    /// one-to-one mapping between identifiers and polynomials.
+    pub(crate) fn register_poly_as(
+        &mut self,
+        id: Id,
+        poly: &Polynomial<F, B>,
+    ) -> AstLeaf<E, Id, B> {
+        self.polys.entry(id).or_insert_with(|| poly.clone());
```

Comment:
Consider treating reuse of the same id as an error.

Reply:
The point of having the ID specified was to avoid a more costly polynomial equality check, while being able to ensure that registering the "same" polynomial resulted in the same `AstLeaf`. But I guess we still get the benefit of the `HashMap` removing the O(n) registration time, so I can add this check and then we can always remove it later if it becomes a bottleneck.

Reply:
I've removed the ID type parameter, as we figured out how to avoid it in a pairing.